provider/oci/storage_volumes.go

```diff
@@ -4,43 +4,567 @@
+	request := ociCore.GetVolumeRequest{
+		VolumeId: resourceID,
+	}
+
+	response, err := v.storageAPI.GetVolume(context.Background(), request)
+	if err != nil {
+		if v.env.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("volume not found: %s", *resourceID)
+		} else {
+			return "", err
+		}
+	}
+	return string(response.Volume.LifecycleState), nil
+}
+
+func (v *volumeSource) createVolume(p storage.VolumeParams, instanceMap map[instance.Id]*ociInstance) (_ *storage.Volume, err error) {
+	var details ociCore.CreateVolumeResponse
+	defer func() {
+		if err != nil && details.Id != nil {
+			req := ociCore.DeleteVolumeRequest{
+				VolumeId: details.Id,
+			}
+			response, nestedErr := v.storageAPI.DeleteVolume(context.Background(), req)
+			if nestedErr != nil && !v.env.isNotFound(response.RawResponse) {
+				logger.Warningf("failed to cleanup volume: %s", *details.Id)
+				return
+			}
+			nestedErr = v.env.waitForResourceStatus(
+				v.getVolumeStatus, details.Id,
+				string(ociCore.VolumeLifecycleStateTerminated),
+				5*time.Minute)
+			if nestedErr != nil && !errors.IsNotFound(nestedErr) {
+				logger.Warningf("failed to cleanup volume: %s", *details.Id)
+				return
+			}
+		}
+	}()
+	if err := v.ValidateVolumeParams(p); err != nil {
+		return nil, errors.Trace(err)
+	}
+	if p.Attachment == nil {
+		return nil, errors.Errorf("volume %s has no attachments", p.Tag.String())
+	}
+	instanceId := p.Attachment.InstanceId
+	instance, ok := instanceMap[instanceId]
+	if !ok {
+		ociInstances, err := v.env.getOciInstances(instanceId)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		instance = ociInstances[0]
+		instanceMap[instanceId] = instance
+	}
+
+	availabilityZone := instance.availabilityZone()
+	name := p.Tag.String()
+
+	volTags := map[string]string{}
+	if p.ResourceTags != nil {
+		volTags = p.ResourceTags
+	}
+	volTags[tags.JujuModel] = v.modelUUID
+
+	size := int(p.Size)
+	requestDetails := ociCore.CreateVolumeDetails{
+		AvailabilityDomain: &availabilityZone,
+		CompartmentId:      v.env.ecfg().compartmentID(),
+		DisplayName:        &name,
+		SizeInMBs:          &size,
+		FreeformTags:       volTags,
+	}
+
+	request := ociCore.CreateVolumeRequest{
+		CreateVolumeDetails: requestDetails,
+	}
+
+	result, err := v.storageAPI.CreateVolume(context.Background(), request)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	err = v.env.waitForResourceStatus(
+		v.getVolumeStatus, result.Volume.Id,
+		string(ociCore.VolumeLifecycleStateAvailable),
+		5*time.Minute)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	volumeDetails, err := v.storageAPI.GetVolume(
+		context.Background(), ociCore.GetVolumeRequest{VolumeId: result.Volume.Id})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &storage.Volume{p.Tag, makeVolumeInfo(volumeDetails.Volume)}, nil
+}
+
+func makeVolumeInfo(vol ociCore.Volume) storage.VolumeInfo {
+	var size uint64
+	if vol.SizeInMBs != nil {
```

Comment:
The response can have either? or it always has both, or ? Is there a reason to prefer MB? (presumably just better fidelity?)